### ISSUE TITLE
XL:  TestGetObjectReadSeekFunctional concurent test fail

### ISSUE THREAD

**Issue by hackintoshrao** [minio/minio#2073]
## Expected behaviour

Test to not to freeze. 
## Actual behaviour

Server inactivity and Frozen on Delete Object.
[Here is the block profile](https://github.com/minio/minio/files/344909/block-lok.pdf)
## Steps to reproduce the behaviour

./exec-concurrent TestGetObjectReadSeekFunctional
## Minio version

latest master 
## System information

Minio 3 server

**Comment by hackintoshrao** [comment]
This seems like there's a locking issue.

**Comment by krishnasrinivas** [comment]
@hackintoshrao  is this the source of exec-current.go:

```
package main

import (
    "fmt"
    "os"
    "os/exec"
    "sync"
)

func init() {
    os.Setenv("ACCESS_KEY", "YRDRWWQLEWS9OBJ31GZ2")
    os.Setenv("SECRET_KEY", "y2sSWzx5ytwvkELcxOuSaQ8n3doNqoIilRpb5Kjj")
    os.Setenv("ENDPOINT", "192.168.1.12:9001")
}

func main() {
    var wg sync.WaitGroup
    f, _ := os.Create("output.log")
    defer f.Close()
    testCmd := "./minio.test -test.timeout 3600s"
    if len(os.Args[1]) != 0 {
        testCmd = fmt.Sprintf("./minio.test -test.timeout 3600s -test.run %s", os.Args[1])
    }
    for i := 0; i < 500; i++ {
        wg.Add(1)
        go func(routineId int) {
            defer wg.Done()

            out, err := exec.Command("sh", "-c", testCmd).Output()
            if err != nil {
                fmt.Println(err.Error())
            }
            fmt.Fprintf(f, "\nGoroutine: %d\n", routineId+1)
            fmt.Fprintf(f, string(out))

        }(i)
    }
    wg.Wait()

}
```

**Comment by hackintoshrao** [comment]
@krishnasrinivas Yes. But you gotto get the binary of the minio `api-functional-v4 test` by running `go test -c api-functional-v4_test.go`. I have the setup ready on minio3.

**Comment by hackintoshrao** [comment]
Here is the summary of the detailed investigation. 

The dealock happens only when the read is done after a seek, reading without a seek doesn't cause a deadlock. Here is the minimal code to reproduce the issue. 

```
// Tests get object ReaderSeeker interface methods.
func TestGetObjectReadSeekFunctional(t *testing.T) {
    if testing.Short() {
        t.Skip("skipping functional tests for short runs")
    }

    // Seed random based on current time.
    rand.Seed(time.Now().Unix())

    // Instantiate new minio client object.
    c, err := minio.New(
        os.Getenv("ENDPOINT"),
        os.Getenv("ACCESS_KEY"),
        os.Getenv("SECRET_KEY"),
        false,
    )
    if err != nil {
        t.Fatal("Error:", err)
    }

    // Enable tracing, write to stderr.
    // c.TraceOn(os.Stderr)

    // Set user agent.
    c.SetAppInfo("Minio-go-FunctionalTest", "0.1.0")

    // Generate a new random bucket name.
    bucketName := randString(60, rand.NewSource(time.Now().UnixNano()))

    // Make a new bucket.
    err = c.MakeBucket(bucketName, "us-east-1")
    if err != nil {
        t.Fatal("Error:", err, bucketName)
    }

    buf := bytes.Repeat([]byte("0123456789abcdef"), rand.Intn(1<<20)+32*1024)

    // Save the data
    objectName := randString(60, rand.NewSource(time.Now().UnixNano()))
    n, err := c.PutObject(bucketName, objectName, bytes.NewReader(buf), "binary/octet-stream")
    if err != nil {
        t.Fatal("Error:", err, bucketName, objectName)
    }

    if n != int64(len(buf)) {
        t.Fatalf("Error: number of bytes does not match, want %v, got %v\n", len(buf), n)
    }

    // Read the data back
    r, err := c.GetObject(bucketName, objectName)
    if err != nil {
        t.Fatal("Error:", err, bucketName, objectName)
    }

    /*  st, err := r.Stat()
        if err != nil {
            t.Fatal("Error:", err, bucketName, objectName)
        }
        if st.Size != int64(len(buf)) {
            t.Fatalf("Error: number of bytes in stat does not match, want %v, got %v\n",
                len(buf), st.Size)
        }
    */
    offset := int64(2048)
    n, err = r.Seek(offset, 0)
    if err != nil {
        t.Fatal("Error:", err, offset)
    }
    if n != offset {
        t.Fatalf("Error: number of bytes seeked does not match, want %v, got %v\n",
            offset, n)
    }
    /*  n, err = r.Seek(0, 1)
        if err != nil {
            t.Fatal("Error:", err)
        }
        if n != offset {
            t.Fatalf("Error: number of current seek does not match, want %v, got %v\n",
                offset, n)
        }
        _, err = r.Seek(offset, 2)
        if err == nil {
            t.Fatal("Error: seek on positive offset for whence '2' should error out")
        }
    */
    n, err = r.Seek(-offset, 2)
    if err != nil {
        t.Fatal("Error:", err)
    }
    if n != 0 {
        t.Fatalf("Error: number of bytes seeked back does not match, want 0, got %v\n", n)
    }

    var buffer1 bytes.Buffer
    if n, err = io.CopyN(&buffer1, r, int64(len(buf))); err != nil {
        if err != io.EOF {
            t.Fatal("Error:", err)
        }
    }
    if !bytes.Equal(buf, buffer1.Bytes()) {
        t.Fatal("Error: Incorrect read bytes v/s original buffer.")
    }

    /*  // Seek again and read again.
        n, err = r.Seek(offset-1, 0)
        if err != nil {
            t.Fatal("Error:", err)
        }
        if n != (offset - 1) {
            t.Fatalf("Error: number of bytes seeked back does not match, want %v, got %v\n", offset-1, n)
        }

        var buffer2 bytes.Buffer
        if _, err = io.CopyN(&buffer2, r, st.Size); err != nil {
            if err != io.EOF {
                t.Fatal("Error:", err)
            }
        }
        // Verify now lesser bytes.
        if !bytes.Equal(buf[2047:], buffer2.Bytes()) {
            t.Fatal("Error: Incorrect read bytes v/s original buffer.")
        }*/
    fmt.Println("Before Delete1 ", bucketName, objectName)
    // close the reader.
    r.Close()
    err = c.RemoveObject(bucketName, objectName)
    if err != nil {
        t.Fatal("Error: ", err)
    }
    fmt.Println("Befor4e Delete2")
    err = c.RemoveBucket(bucketName)
    if err != nil {
        t.Fatal("Error:", err)
    }
    fmt.Println("After Delete2")
}
```

Since this bug has been occurring often after changes to GetObject, will be working on replicating it using `server_test` functional test and making it part of the test suite so that any regressions will be caught by travis. 

[July 6]

**Comment by hackintoshrao** [comment]
Now getting data sanctity error too along the lock issue after updating the code to latest master. 
`api-functional-v4_test.go:748: Error: Incorrect read bytes v/s original buffer.`

**Comment by harshavardhana** [comment]
> Now getting data sanctity error too along the lock issue after updating the code to latest master. 
> api-functional-v4_test.go:748: Error: Incorrect read bytes v/s original buffer.

A Seek() here is nothing bug a GetObject Range request, can you paste `.TraceOn` output as well.

**Comment by hackintoshrao** [comment]
Its working fine when  MINIO_CACHE_SIZE is set to 0.

**Comment by harshavardhana** [comment]
Fixed by #2104
